### ISSUE TITLE
Rewrite requests asynchronously

### ISSUE THREAD

**Issue by jsha** [EFForg/https-everywhere#2612]
Right now we do ruleset rewriting synchronously, which is bad because sometimes we need to load a ruleset from disk. Then if we block on I/O, we block the whole ball of wax. We've gotten feedback from Mozilla that this needs to be fixed before our next version or we will get rejected from AMO (though I think our self-hosted version is fine).

Instead, we can suspend an nsIRequest and resume it in the callback from our I/O. IRC transcript:

```
< jsha> I'm looking for advice on redirecting HTTP requests asynchronously. Andreas suggested there is a way?
< jsha> In particular, HTTPS Everywhere potentially needs to do some file I/O in response to an HTTP request (read a ruleset off disk), and I'd like to be able to say
        "Don't let this request proceed until I come back with a callback."
< jorgev> The_8472: ^^
< The_8472> jsha http-on-modify-request should give you an https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsIHttpChannel
< freonbot> [ nsIHttpChannel - Mozilla | MDN ] - developer.mozilla.org
< The_8472> just suspend it in the observer notification and resume it in your callback
< jsha> How do I suspend it?
< The_8472> but idk if that's reentrant. if other addons do the same that might be a problem
< The_8472> look at its superinterfaces
< jsha> aha, it's on nsIRequest. Thanks, I'll try that.
< jsha> How do I check if it's reentrant?
< Piccaruse_> Hey, would it be possible to replace an element on a page (ie a flashplayer) with a vlc player through an extension?
< The_8472> jsha, i guess register two observers and see if the second one gets called after a suspend. if no further observers get notified until the resume then it'll
            probably avoid most races
```

**Comment by jsha** [comment]
Turns out this doesn't work in the straightforward way you would hope, but Giorgio of Noscript has found a workaround: https://bugzilla.mozilla.org/show_bug.cgi?id=1170197.

**Comment by jsha** [comment]
Branch in progress at https://github.com/EFForg/https-everywhere/compare/async?expand=1 mostly works.

**Comment by jsha** [comment]
Branch updated now to include cookie rewriting and initial rule loading. Tests currently fail though.

**Comment by gloomy-ghost** [comment]
Closed by #4110
